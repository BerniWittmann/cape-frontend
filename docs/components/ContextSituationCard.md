# ContextSituationCard

A Card to view and edit a single Context Situation

## Props

<!-- @vuese:ContextSituationCard:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|contextSituation|The ContextSituation Object|`Object`|`true`|-|

<!-- @vuese:ContextSituationCard:props:end -->


## Events

<!-- @vuese:ContextSituationCard:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|deactivate|Fired when the Card should be closed|-|

<!-- @vuese:ContextSituationCard:events:end -->


