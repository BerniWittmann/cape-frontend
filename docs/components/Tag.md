# Tag

A Component to view a single Tag

## Props

<!-- @vuese:Tag:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|tag|The Tag object|`Object`|`true`|-|
|closable|Option whether the tag should be deleteable or not|`Boolean`|`false`|false|
|size|Size of the Tag|`'medium'` / `'small'` / `'mini'`|`false`|-|

<!-- @vuese:Tag:props:end -->


## Events

<!-- @vuese:Tag:events:start -->
|Event Name|Description|Parameters|
|---|---|---|
|close|Emitted when the close button is clicked|-|

<!-- @vuese:Tag:events:end -->


