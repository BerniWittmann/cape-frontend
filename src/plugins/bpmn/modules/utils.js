import {
  NO_CONNECTIONS_DISALLOWED_OBJECTS,
  END_EVENT_TYPE,
  START_EVENT_TYPE,
  MULTIPLE_SEQUENCE_FLOW_ALLOWED_OBJECT_TYPES,
  CONNECTION_RULE_ENFORCED_ELEMENTS,
  SEQUENCE_FLOW_TYPE
} from './constants'

export function shapeIsType(shape, type) {
  if (!shape || !shape.businessObject) return false
  return shape.businessObject.$type === type
}

export function shapeHasType(shape, types) {
  if (!shape || !shape.businessObject) return false
  return types.includes(shape.businessObject.$type)
}

export function getElementsByType(elementRegistry, type) {
  return elementRegistry.filter(el => shapeIsType(el, type))
}

export function hasUnconnectedActivities(elementRegistry) {
  const unconnectedElements = elementRegistry.filter(el => {
    if (!shapeHasType(el, NO_CONNECTIONS_DISALLOWED_OBJECTS)) return false
    if (el.type === 'label') return false
    const hasIncomingConnections = el.incoming.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length > 0
    const hasOutgoingConnections = el.outgoing.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length > 0
    if (shapeIsType(el, END_EVENT_TYPE) && hasIncomingConnections) return false
    if (shapeIsType(el, START_EVENT_TYPE) && hasOutgoingConnections) return false
    return !(hasIncomingConnections && hasOutgoingConnections)
  })
  return unconnectedElements.length > 0
}

export function hasExceedingConnectionCountElements(elementRegistry) {
  const invalidConnectedElements = elementRegistry.filter(el => {
    if (shapeHasType(el, MULTIPLE_SEQUENCE_FLOW_ALLOWED_OBJECT_TYPES)) return false
    if (!shapeHasType(el, CONNECTION_RULE_ENFORCED_ELEMENTS)) return false
    const incomingConnections = el.incoming.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length
    const outgoingConnections = el.outgoing.filter(c => shapeIsType(c, SEQUENCE_FLOW_TYPE)).length
    if (shapeIsType(el, END_EVENT_TYPE) && incomingConnections === 1) return false
    if (shapeIsType(el, START_EVENT_TYPE) && outgoingConnections === 1) return false
    return !(incomingConnections === 1 && outgoingConnections === 1)
  })
  return invalidConnectedElements.length > 0
}

export function getStartEventCount(elementRegistry) {
  const startEvents = elementRegistry.filter(el => {
    return shapeIsType(el, START_EVENT_TYPE)
  })
  return startEvents.length
}

export function getEndEventCount(elementRegistry) {
  const endEvents = elementRegistry.filter(el => {
    return shapeIsType(el, END_EVENT_TYPE)
  })
  return endEvents.length
}
