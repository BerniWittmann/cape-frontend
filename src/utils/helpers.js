export function convertHexToRgba(color, opacity = 1) {
  const c = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(color)
  return `rgba(${parseInt(c[1], 16)}, ${parseInt(c[2], 16)}, ${parseInt(c[3], 16)}, ${opacity})`
}

export function getCookie(name) {
  const value = '; ' + document.cookie
  const parts = value.split('; ' + name + '=')
  if (parts.length === 2) return parts.pop().split(';').shift()
}

export function hasProcessModelerRulesEnabled() {
  const cookie = getCookie('enableProcessValidation')
  return cookie ? cookie === 'true' : true
}

export function removeByID(arr, id) {
  return arr.filter(o => o.id !== id)
}

export function updateAndSetActive(store, data, Model, moduleName) {
  data = new Model(data)
  store.dispatch(`${moduleName}/update`, data).then(() => {
    store.dispatch(`${moduleName}/setActive`, data)
  })
}
