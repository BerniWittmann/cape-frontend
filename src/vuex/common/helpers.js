import Vue from 'vue'

function pushToChildren(tagMap, id, data) {
  const tag = tagMap.get(id)
  if (!tag) return
  tag.children.push(data)
  tagMap.set(id, tag)
}

export function update(state, key, activeKey, data) {
  state[key] = [...state[key].map((p) => p.id === data.id ? { ...data } : p)]
  if (state[activeKey] && state[activeKey].id === data.id) {
    state[activeKey] = { ...data }
  }
}

export function setActive(state, key, activeKey, data) {
  if (!data || !data.id) {
    state[activeKey] = undefined
  } else {
    state[activeKey] = state[key].find((p) => p.id === data.id)
  }
}

export function getGraphNodes(objects, type, routeName, routeID) {
  return objects.map((o) => ({
    id: o.id,
    name: o.name,
    type: type,
    route: {
      name: routeName,
      params: {
        [routeID]: o.id
      }
    }
  }))
}

export function getByTags(objects, tags) {
  const tagMap = new Map()
  tagMap.set('untagged', {
    value: '0',
    label: Vue.i18n.t('context_situation.untagged'),
    children: []
  })
  tags.forEach(tag => {
    tagMap.set(tag.id, {
      value: tag.id,
      label: tag.name,
      children: []
    })
  })
  objects.forEach((obj) => {
    const objData = {
      value: obj,
      key: obj.id,
      label: obj.name
    }
    if (obj.tags && obj.tags.length > 0) {
      obj.tags.forEach((tag) => {
        pushToChildren(tagMap, tag.id, objData)
      })
    } else {
      pushToChildren(tagMap, 'untagged', objData)
    }
  })
  return [...tagMap.values()]
}
