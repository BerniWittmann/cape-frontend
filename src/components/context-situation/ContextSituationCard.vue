<template>
  <el-card :class="{ 'context-situation-card': true, 'is-active': isActive }"
           @click.native.stop="navigateToContextSituation" :shadow="shadowMode">
    <div slot="header" class="clearfix">
      <input-edit
              :value="contextSituationData.name" :rules="rules.name"
              :placeholder="$t('context_situation.edit.name')"
      ></input-edit>
    </div>
    <div class="tags" v-if="isActive">
        <tag-editor v-model="contextSituationData.tags" @change="() => {}"></tag-editor>
    </div>
    <div class="tags" v-else>
      <tag v-for="tag in contextSituation.tags" :tag="tag" size="medium" :key="tag.id"></tag>
    </div>

    <ContextSituationRules v-if="isActive" :context-situation="contextSituationData"
                           class="top-space"></ContextSituationRules>

    <div v-if="isActive" class="top-space">
      <el-button type="danger" plain @click="resetCS" size="medium">
        {{ $t('context_situation.edit.reset') }}
      </el-button>
      <el-button type="danger" @click="deleteCS" size="medium">
        {{ $t('context_situation.edit.delete') }}
      </el-button>
      <el-button style="float: right" type="success" @click="saveCS" size="medium">{{
        $t('context_situation.edit.save') }}
      </el-button>
    </div>
  </el-card>
</template>

<script>
/* ============
 * Context Situation Card Component
 * ============
 *
 * A Card to view a single Context Situation
 */
import Tag from '@/components/Tag'
import TagEditor from '@/components/TagEditor'
import ContextSituationRules from './ContextSituationRules'
import ContextSituationService from '@/services/contextSituation'
import InputEdit from '@/components/InputEdit.vue'

export default {
  components: {
    ContextSituationRules,
    Tag,
    TagEditor,
    InputEdit
  },

  props: {
    contextSituation: {
      type: Object,
      required: true
    }
  },

  data() {
    return {
      contextSituationData: {},
      rules: {
        name: [
          { required: true, message: this.$t('context_situation.edit.validation.name.required'), trigger: 'blur' }
        ]
      }
    }
  },

  computed: {
    activeContextSituation() {
      return this.$store.state.contextSituation.activeContextSituation
    },

    isActive() {
      return this.activeContextSituation && this.activeContextSituation.id === this.contextSituation.id
    },

    shadowMode() {
      return this.isActive ? 'always' : 'hover'
    }
  },

  methods: {
    navigateToContextSituation() {
      if (!this.isActive) {
        this.contextSituationData = this.copyContextSituation(this.contextSituation)
        this.$router.push({
          name: 'context_situations.single',
          params: {
            contextSituationID: this.contextSituation.id
          }
        })
      }
    },

    saveCS() {
      // validation is done because of the way the input works, need to be updated if rules are no input anymore
      const data = { ...this.contextSituationData }
      ContextSituationService.update(data).then(result => {
        if (result) {
          this.$emit('deactivate')
        }
      })
    },

    resetCS() {
      this.contextSituationData = { ...this.contextSituation }
    },

    deleteCS() {
      this.$confirm(this.$t('context_situation.delete.message'), this.$t('context_situation.delete.warning'), {
        confirmButtonText: this.$t('context_situation.delete.ok'),
        cancelButtonText: this.$t('context_situation.delete.cancel'),
        type: 'warning',
        cancelButtonClass: 'is-plain el-button--info',
        confirmButtonClass: 'el-button--danger'
      }).then(() => {
        ContextSituationService.remove(this.activeContextSituation).then(() => {
          this.$router.back()
          ContextSituationService.getAll()
        })
      }).catch(() => {
        this.$message({
          type: 'info',
          message: this.$t('context_situation.delete.cancellation')
        })
      })
    },

    copyContextSituation(csIn) {
      return {
        id: csIn.id,
        name: csIn.name,
        tags: csIn.tags.slice(),
        rules: csIn.rules
      }
    }
  },

  beforeMount() {
    this.contextSituationData = this.copyContextSituation(this.contextSituation)
  }
}

</script>

<style lang="scss" scoped>
.context-situation-card {
  margin: 10px;
  width: 300px;
  cursor: pointer;

  &.is-active {
    width: 100%;
    margin-bottom: 50px;
  }
}

.tag-space {
  margin-left: 5px;
}

.top-space {
  margin-top: 20px;
}

.title {
  font-size: large;
  font-weight: bold;
}

.el-form-item {
  margin-top: -10px;
  margin-bottom: -10px;
}

</style>
