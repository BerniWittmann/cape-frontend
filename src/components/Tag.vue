<template>
  <el-tag class="tag" type="info" :style="style" :closable="closable" :size="size" @close="emitClose">{{ tag.name }}</el-tag>
</template>

<script>
import { convertHexToRgba } from '@/utils/helpers'

/*
 * @vuese
 * @group Components
 *
 * A Component to view a single Tag
 */
export default {
  name: 'Tag',
  props: {
    // The Tag object
    tag: {
      type: Object,
      required: true
    },

    // Option whether the tag should be deleteable or not
    closable: {
      type: Boolean,
      default: false
    },

    // Size of the Tag
    size: {
      // `'medium'` / `'small'` / `'mini'`
      type: String
    }
  },

  computed: {
    style() {
      if (!this.tag.color) return {}

      return {
        backgroundColor: convertHexToRgba(this.tag.color, 0.1),
        borderColor: convertHexToRgba(this.tag.color, 0.2),
        color: convertHexToRgba(this.tag.color, 1)
      }
    }
  },

  methods: {
    emitClose() {
      // Emitted when the close button is clicked
      this.$emit('close')
    }
  }
}

</script>

<style lang="scss">
.tag {
  margin-right: 5px;
  &:last-of-type {
    margin-right: 0;
  }

  .el-tag__close {
    color: inherit !important;
    &:hover {
      background-color: inherit !important;
      color: #fff;
    }
  }
}
</style>
