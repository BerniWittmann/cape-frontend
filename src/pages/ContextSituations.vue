<template>
  <v-layout>
    <el-row justify="end" :gutter="20">
      <el-col :span="16">
        <el-form ref='newFactorForm' :model="newContextSituation" label-position="top" :rules="rules" status-icon inline
                 @submit.native.prevent>
          <el-form-item prop="name">
            <el-input
                    :placeholder="$t('context_situation.new_context_situation_name')"
                    v-model="newContextSituation.name">
            </el-input>
          </el-form-item>
          <el-button icon="el-icon-plus" @click="createNew">{{ $t('context_situation.create_new') }}</el-button>
        </el-form>
      </el-col>
      <el-col :span="3">
        <el-select v-if="tags.length > 0"
                :placeholder="$t('context_situation.select_placeholder')"
                v-model="selectedTags" clearable multiple collapse-tags>
          <i slot="prefix" class="el-input__icon el-icon-view"></i>
          <el-option v-for="tag in tags" :key="tag.id" :label="tag.name" :value="tag.id">
            <Tag :tag="tag" size="mini"></Tag>
          </el-option>
        </el-select>
      </el-col>
      <el-col :span="5">
        <el-input
                :placeholder="$t('context_situation.filter_placeholder')"
                v-model="filterText" clearable>
          <i slot="prefix" class="el-input__icon el-icon-search"></i>
        </el-input>
      </el-col>
    </el-row>
    <el-row :gutter="20" @click.native="unsetActiveContextSituation" ref="context-situation-cards">
      <el-col :span="24">
        <transition-group tag="div" name="context-situation-cards" class="context-situation-cards">
          <context-situation-card v-for="situation in contextSituations"
                                  :key="situation.id"
                                  :context-situation="situation"
                                  v-on:deactivate="unsetActiveContextSituation"></context-situation-card>
        </transition-group>
        <transition name="el-fade-in">
          <p v-if="contextSituations.length === 0">{{ $t('context_situation.none_found') }}</p>
        </transition>
      </el-col>
    </el-row>
  </v-layout>
</template>

<script>
import { scrollToTop, hasCommonElement } from '@/utils/helpers'
import DefaultLayout from '@/layouts/Default.vue'
import ContextSituationCard from '@/components/context-situation/ContextSituationCard.vue'
import ContextSituationService from '@/services/contextSituation'
import Tag from '@/components/Tag'
import ContextSituation from '@/models/contextSituation'

export default {
  components: {
    VLayout: DefaultLayout,
    ContextSituationCard,
    Tag
  },

  data() {
    return {
      filterText: undefined,
      selectedTags: [],
      newContextSituation: {
        name: undefined
      },
      rules: {
        name: [
          { required: true, message: this.$t('context_situation.edit.validation.name.required'), trigger: 'blur' }
        ]
      }
    }
  },

  computed: {
    activeContextSituation() {
      return this.$store.state.contextSituation.activeContextSituation
    },

    allContextSituations() {
      return this.$store.state.contextSituation.contextSituations
    },

    filteredContextSituations() {
      if (!this.filterText && !this.selectedTags) return this.allContextSituations

      return this.allContextSituations.filter((current) => {
        if (this.activeContextSituation && current.id === this.activeContextSituation.id) return true

        if (this.selectedTags.length > 0) {
          const tagIDs = current.tags.map((t) => t.id)
          if (!hasCommonElement(tagIDs, this.selectedTags)) return false
        }

        if (!this.filterText) return true
        return current.name.toLowerCase().includes(this.filterText.toLowerCase())
      })
    },

    contextSituations() {
      return [...this.filteredContextSituations].sort((a, b) => {
        if (this.activeContextSituation) {
          if (a.id === this.activeContextSituation.id) return -1
          if (b.id === this.activeContextSituation.id) return 1
        }

        return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1
      })
    },

    tags() {
      return this.$store.state.tag.tags
    }
  },

  methods: {
    createNew() {
      this.$refs.newFactorForm.validate((valid) => {
        if (valid) {
          let data = new ContextSituation({ name: this.newContextSituation.name })
          ContextSituationService.create(data).then(() => {
            this.$router.push({
              name: 'context_situations.single',
              params: {
                contextSituationID: this.$store.state.contextSituation.contextSituations[this.$store.state.contextSituation.contextSituations.length - 1].id
              }
            })
          })
        }
      })
      this.$refs.newFactorForm.resetFields()
    },

    unsetActiveContextSituation() {
      this.$store.dispatch('contextSituation/unsetActive')
      this.$router.push({
        name: 'context_situations'
      })
    }
  },

  beforeRouteUpdate(to, from, next) {
    if (to.name === 'context_situations.single') {
      scrollToTop()
      ContextSituationService.get({ id: to.params.contextSituationID }).then(next)
    } else {
      this.$store.dispatch('contextSituation/unsetActive').then(next)
    }
  },

  beforeRouteLeave(to, from, next) {
    this.$store.dispatch('contextSituation/unsetActive').then(next)
  },

  beforeRouteEnter(to, from, next) {
    scrollToTop()
    next()
  }
}

</script>

<style lang="scss">
.context-situation-cards {
  margin-top: 20px;
  overflow: hidden;
  display: flex;
  flex-wrap: wrap;
  align-content: space-around;
}

.context-situation-cards-move {
  transition: transform 0.500s ease-in-out !important;
}
</style>
