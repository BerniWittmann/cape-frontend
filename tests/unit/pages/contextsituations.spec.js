import { shallowMount } from '@vue/test-utils'
import { i18n } from '../setupPlugins'

import Tag from '@/components/Tag.vue'
import ContextSituationCard from '@/components/context-situation/ContextSituationCard.vue'
import ContextSituations from '@/pages/ContextSituations.vue'
import ContextSituationService from '@/services/contextSituation'

import * as helpers from '@/utils/helpers'

jest.mock('@/utils/helpers')

const situations = [{
  id: 1,
  name: 'CS A',
  tags: [{
    id: 101,
    name: 'Tag 1',
    color: '#FF0000'
  }]
}, {
  id: 2,
  name: 'CS F',
  tags: [{
    id: 101,
    name: 'Tag 1',
    color: '#FF0000'
  }, {
    id: 102,
    name: 'Tag 2',
    color: '#FFF000'
  }]
}, {
  id: 3,
  name: 'CS C',
  tags: [{
    id: 102,
    name: 'Tag 2',
    color: '#FFF000'
  }, {
    id: 103,
    name: 'Tag 3',
    color: '#00FFF0'
  }]
}, {
  id: 4,
  name: 'CS D',
  tags: []
}, {
  id: 5,
  name: 'CS E',
  tags: []
}, {
  id: 6,
  name: 'CS B',
  tags: [{
    id: 101,
    name: 'Tag 1',
    color: '#FF0000'
  }, {
    id: 103,
    name: 'Tag 3',
    color: '#00FFF0'
  }]
}]

describe('Pages', () => {
  describe('ContextSituations.vue', () => {
    let store
    let router
    let cmp
    ContextSituationService.get = jest.fn().mockReturnValue({ then: (arg) => arg() })

    beforeEach(() => {
      router = {
        push: jest.fn()
      }
      store = {
        state: {
          tag: {
            tags: [{
              id: 101,
              name: 'Tag 1',
              color: '#FF0000'
            }, {
              id: 102,
              name: 'Tag 2',
              color: '#FFF000'
            }, {
              id: 103,
              name: 'Tag 3',
              color: '#00FFF0'
            }, {
              id: 104,
              name: 'Tag 4',
              color: '#000000'
            }]
          },
          contextSituation: {
            activeContextSituation: undefined,
            contextSituations: situations
          }
        },
        dispatch: jest.fn().mockReturnValue({ then: (arg) => arg() })
      }
      render()
    })

    function render() {
      cmp = shallowMount(ContextSituations, {
        i18n,
        mocks: {
          $store: store,
          $router: router
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders the Situations', () => {
      const situations = cmp.findAll(ContextSituationCard)

      expect(situations.length).toEqual(6)

      expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[0])
      expect(situations.at(1).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[5])
      expect(situations.at(2).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[2])
      expect(situations.at(3).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[3])
      expect(situations.at(4).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[4])
      expect(situations.at(5).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[1])
    })

    describe('renders the active Situation first', () => {
      for (let i = 0; i < situations.length; i++) {
        it('holds for situation ' + i, () => {
          store.state.contextSituation.activeContextSituation = store.state.contextSituation.contextSituations[i]

          const situations = cmp.findAll(ContextSituationCard)

          expect(situations.length).toEqual(6)

          expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[i])
        })
      }
    })

    describe('it can filter by tags', () => {
      it('renders a select for the tags', () => {
        const sel = cmp.find('elselect-stub')

        expect(sel.exists()).toBeTruthy()
        expect(sel.attributes('clearable')).toEqual('true')

        const opts = sel.findAll('eloption-stub')
        expect(opts.length).toEqual(4)

        for (let i = 0; i < opts.length; i++) {
          const opt = opts.at(i)
          const tag = store.state.tag.tags[i]

          expect(opt.contains(Tag)).toBeTruthy()
          expect(opt.find(Tag).props('tag')).toEqual(tag)
          expect(opt.attributes('label')).toEqual(tag.name)
          expect(opt.attributes('value')).toEqual(tag.id.toString())
        }
      })

      it('does not render the select field if no tags available', () => {
        store.state.tag.tags = []
        render()
        expect(cmp.contains('elselect-stub')).toBeFalsy()
      })

      it('only displays situations with the correct tags', () => {
        cmp.vm.selectedTags = [101]
        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(3)

        expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[0])
        expect(situations.at(1).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[5])
        expect(situations.at(2).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[1])
      })

      it('can select multiple tags', () => {
        const sel = cmp.find('elselect-stub')

        expect(sel.exists()).toBeTruthy()
        expect(sel.attributes('multiple')).toEqual('true')

        cmp.vm.selectedTags = [101, 103]

        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(4)

        expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[0])
        expect(situations.at(1).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[5])
        expect(situations.at(2).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[2])
        expect(situations.at(3).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[1])
      })

      it('displays message with none found', () => {
        cmp.vm.selectedTags = [104]
        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(0)

        const text = cmp.find('p')
        expect(text.exists()).toBeTruthy()
        expect(text.text()).toEqual('context_situation.none_found')
      })

      it('still renders the active Situation although it does not have the correct tag', () => {
        store.state.contextSituation.activeContextSituation = store.state.contextSituation.contextSituations[3]
        const sel = cmp.find('elselect-stub')

        expect(sel.exists()).toBeTruthy()
        expect(sel.attributes('multiple')).toEqual('true')

        cmp.vm.selectedTags = [101, 103]

        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(5)

        expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[3])
        expect(situations.at(1).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[0])
        expect(situations.at(2).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[5])
        expect(situations.at(3).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[2])
        expect(situations.at(4).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[1])
      })
    })
    describe('creating new context Situation', () => {
      it('can create new context Situation', () => {
        cmp.vm.newContextSituation.name = 'New CS'
        cmp.vm.$refs.newFactorForm.resetFields = jest.fn()
        cmp.vm.$refs.newFactorForm.validate = jest.fn().mockImplementation((valid) => {
          valid(true)
        })
        ContextSituationService.create = jest.fn().mockImplementation(() => ({
          then: (arg) => arg()
        }))
        ContextSituationService.getAll = jest.fn()
        cmp.vm.createNew()
        expect(ContextSituationService.create).toHaveBeenCalledWith({
          '_id': undefined,
          'tags': [],
          'rules': undefined,
          'name': 'New CS'
        })
      })

      it('cannot create new context Situation if validation failed', () => {
        cmp.vm.newContextSituation.name = 'New CS'
        cmp.vm.$refs.newFactorForm.validate = jest.fn().mockImplementation((valid) => {
          valid(false)
        })
        ContextSituationService.create = jest.fn().mockImplementation(() => ({
          then: (arg) => arg()
        }))
        cmp.vm.$refs.newFactorForm.resetFields = jest.fn()
        cmp.vm.createNew()
        expect(ContextSituationService.create).not.toHaveBeenCalled()
      })
    })

    describe('it can filter by search Text', () => {
      it('renders a input for the text', () => {
        const input = cmp.findAll('elinput-stub').at(1)
        expect(input.exists()).toBeTruthy()
        expect(input.attributes('clearable')).toEqual('true')
      })

      it('only displays situations with the correct name', () => {
        cmp.vm.filterText = 'b'
        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(1)

        expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[5])
      })

      it('displays all situations with no name', () => {
        cmp.vm.filterText = ''
        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(6)
      })

      it('displays message with none found', () => {
        cmp.vm.filterText = 'arbitrary so that it does not find'
        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(0)

        const text = cmp.find('p')
        expect(text.exists()).toBeTruthy()
        expect(text.text()).toEqual('context_situation.none_found')
      })

      it('still renders the active Situation although it does not have the correct name', () => {
        store.state.contextSituation.activeContextSituation = store.state.contextSituation.contextSituations[3]
        const sel = cmp.find('elselect-stub')

        expect(sel.exists()).toBeTruthy()
        expect(sel.attributes('multiple')).toEqual('true')

        cmp.vm.filterText = 'A'

        const situations = cmp.findAll(ContextSituationCard)

        expect(situations.length).toEqual(2)

        expect(situations.at(0).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[3])
        expect(situations.at(1).props('contextSituation')).toEqual(store.state.contextSituation.contextSituations[0])
      })
    })

    it('when clicking the background it unselects the currently active Situation', () => {
      store.state.contextSituation.activeContextSituation = store.state.contextSituation.contextSituations[3]

      const bg = cmp.find({ ref: 'context-situation-cards' })
      bg.trigger('click')

      expect(router.push).toHaveBeenCalledWith({
        name: 'context_situations'
      })
      expect(store.dispatch).toHaveBeenCalledWith('contextSituation/unsetActive')
    })

    describe('navigation Guards', () => {
      let next
      beforeEach(() => {
        next = jest.fn()
        jest.spyOn(helpers, 'scrollToTop')
        jest.clearAllMocks()
      })
      describe('beforeRouteEnter', () => {
        it('it has the beforeRouteEnter Hook', () => {
          expect(cmp.vm.$options.beforeRouteEnter).toEqual(expect.any(Function))
        })

        it('scrolls to top', () => {
          cmp.vm.$options.beforeRouteEnter({}, {}, next)

          expect(helpers.scrollToTop).toHaveBeenCalled()
          expect(next).toHaveBeenCalled()
        })
      })

      describe('beforeRouteUpdate', () => {
        it('it has the beforeRouteUpdate Hook', () => {
          expect(cmp.vm.$options.beforeRouteUpdate).toEqual(expect.any(Function))
        })

        it('loads the correct situation when going to the situation route', () => {
          cmp.vm.$options.beforeRouteUpdate.call({ $store: store }, {
            name: 'context_situations.single',
            params: { contextSituationID: '42' }
          }, {}, next)

          expect(ContextSituationService.get).toHaveBeenCalledWith({ id: '42' })
          expect(helpers.scrollToTop).toHaveBeenCalled()
          expect(store.dispatch).not.toHaveBeenCalled()
          expect(next).toHaveBeenCalled()
        })

        it('unsets the active Context Situation for other routes', () => {
          cmp.vm.$options.beforeRouteUpdate.call({ $store: store }, { name: 'other' }, {}, next)

          expect(store.dispatch).toHaveBeenCalledWith('contextSituation/unsetActive')
          expect(helpers.scrollToTop).not.toHaveBeenCalled()
          expect(ContextSituationService.get).not.toHaveBeenCalled()
          expect(next).toHaveBeenCalled()
        })
      })

      describe('beforeRouteLeave', () => {
        it('it has the beforeRouteLeave Hook', () => {
          expect(cmp.vm.$options.beforeRouteLeave).toEqual(expect.any(Function))
        })

        it('unsets the active Context Situation', () => {
          cmp.vm.$options.beforeRouteLeave.call({ $store: store }, {}, {}, next)

          expect(store.dispatch).toHaveBeenCalledWith('contextSituation/unsetActive')
          expect(next).toHaveBeenCalled()
        })
      })
    })
  })
})
