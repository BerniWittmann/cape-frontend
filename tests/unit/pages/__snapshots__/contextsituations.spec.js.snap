// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages ContextSituations.vue renders 1`] = `
<vlayout-stub>
  <elrow-stub tag="div" gutter="20" justify="end" align="top">
    <elcol-stub span="16" tag="div">
      <elform-stub model="[object Object]" rules="[object Object]" labelposition="top" labelsuffix="" inline="true" statusicon="true" showmessage="true" validateonrulechange="true">
        <elformitem-stub prop="name" inlinemessage="" showmessage="true">
          <elinput-stub type="text" autocomplete="off" validateevent="true" placeholder="context_situation.new_context_situation_name"></elinput-stub>
        </elformitem-stub>
        <elbutton-stub type="default" icon="el-icon-plus" nativetype="button">context_situation.create_new</elbutton-stub>
      </elform-stub>
    </elcol-stub>
    <elcol-stub span="3" tag="div">
      <elselect-stub value="" autocomplete="off" clearable="true" multiple="true" multiplelimit="0" placeholder="context_situation.select_placeholder" valuekey="value" collapsetags="true" popperappendtobody="true"><i class="el-input__icon el-icon-view"></i>
        <eloption-stub value="101" label="Tag 1">
          <tag-stub tag="[object Object]" size="mini"></tag-stub>
        </eloption-stub>
        <eloption-stub value="102" label="Tag 2">
          <tag-stub tag="[object Object]" size="mini"></tag-stub>
        </eloption-stub>
        <eloption-stub value="103" label="Tag 3">
          <tag-stub tag="[object Object]" size="mini"></tag-stub>
        </eloption-stub>
        <eloption-stub value="104" label="Tag 4">
          <tag-stub tag="[object Object]" size="mini"></tag-stub>
        </eloption-stub>
      </elselect-stub>
    </elcol-stub>
    <elcol-stub span="5" tag="div">
      <elinput-stub type="text" autocomplete="off" validateevent="true" clearable="true" placeholder="context_situation.filter_placeholder"><i class="el-input__icon el-icon-search"></i></elinput-stub>
    </elcol-stub>
  </elrow-stub>
  <elrow-stub tag="div" gutter="20" justify="start" align="top">
    <elcol-stub span="24" tag="div"><span tag="div" name="context-situation-cards" class="context-situation-cards"><contextsituationcard-stub contextsituation="[object Object]"></contextsituationcard-stub><contextsituationcard-stub contextsituation="[object Object]"></contextsituationcard-stub><contextsituationcard-stub contextsituation="[object Object]"></contextsituationcard-stub><contextsituationcard-stub contextsituation="[object Object]"></contextsituationcard-stub><contextsituationcard-stub contextsituation="[object Object]"></contextsituationcard-stub><contextsituationcard-stub contextsituation="[object Object]"></contextsituationcard-stub></span>
      <!---->
    </elcol-stub>
  </elrow-stub>
</vlayout-stub>
`;
